File: csharp/CsSearch/CsSearch/SearchOptions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Xml.Linq;

namespace CsSearch
{
	public class SearchOptions
	{
		private readonly string _searchOptionsResource;

		public static Dictionary<string, Action<string, SearchSettings>> ArgActionDictionary =
			new Dictionary<string, Action<string,SearchSettings>>
				{
					{ "encoding", (s, settings) => settings.TextFileEncoding = s },
					{ "in-archiveext", (s, settings) => settings.AddInArchiveExtension(s) },
					{ "in-archivefilepattern", (s, settings) => settings.AddInArchiveFilePattern(s) },
					{ "in-dirpattern", (s, settings) => settings.AddInDirPattern(s) },
					{ "in-ext", (s, settings) => settings.AddInExtension(s) },
					{ "in-filepattern", (s, settings) => settings.AddInFilePattern(s) },
					{ "in-filetype", (s, settings) => settings.AddInFileType(s) },
					{ "in-linesafterpattern", (s, settings) => settings.AddInLinesAfterPattern(s) },
					{ "in-linesbeforepattern", (s, settings) => settings.AddInLinesBeforePattern(s) },
					{ "linesafter", (s, settings) => settings.LinesAfter = int.Parse(s) },
					{ "linesaftertopattern", (s, settings) => settings.AddLinesAfterToPattern(s) },
					{ "linesafteruntilpattern", (s, settings) => settings.AddLinesAfterUntilPattern(s) },
					{ "linesbefore", (s, settings) => settings.LinesBefore = int.Parse(s) },
					{ "maxlinelength", (s, settings) => settings.MaxLineLength = int.Parse(s) },
					{ "out-archiveext", (s, settings) => settings.AddOutArchiveExtension(s) },
					{ "out-archivefilepattern", (s, settings) => settings.AddOutArchiveFilePattern(s) },
					{ "out-dirpattern", (s, settings) => settings.AddOutDirPattern(s) },
					{ "out-ext", (s, settings) => settings.AddOutExtension(s) },
					{ "out-filepattern", (s, settings) => settings.AddOutFilePattern(s) },
					{ "out-filetype", (s, settings) => settings.AddOutFileType(s) },
					{ "out-linesafterpattern", (s, settings) => settings.AddOutLinesAfterPattern(s) },
					{ "out-linesbeforepattern", (s, settings) => settings.AddOutLinesBeforePattern(s) },
					{ "search", (s, settings) => settings.AddSearchPattern(s) },
				};

		public static Dictionary<string, Action<bool, SearchSettings>> BoolFlagActionDictionary =
			new Dictionary<string, Action<bool, SearchSettings>>
				{
					{ "allmatches", (b, settings) => settings.FirstMatch = !b },
					{ "archivesonly", (b, settings) => settings.SetArchivesOnly(b) },
					{ "debug", (b, settings) => settings.SetDebug(b) },
					{ "excludehidden", (b, settings) => settings.ExcludeHidden = b },
					{ "firstmatch", (b, settings) => settings.FirstMatch = b },
					{ "help", (b, settings) => settings.PrintUsage = b },
					{ "includehidden", (b, settings) => settings.ExcludeHidden = !b },
					{ "listdirs", (b, settings) => settings.ListDirs = b },
					{ "listfiles", (b, settings) => settings.ListFiles = b },
					{ "listlines", (b, settings) => settings.ListLines = b },
					{ "multilinesearch", (b, settings) => settings.MultiLineSearch = b },
					{ "noprintmatches", (b, settings) => settings.PrintResults = !b },
					{ "norecursive", (b, settings) => settings.Recursive = !b },
					{ "nosearcharchives", (b, settings) => settings.SearchArchives = !b },
					{ "printmatches", (b, settings) => settings.PrintResults = b },
					{ "recursive", (b, settings) => settings.Recursive = b },
					{ "searcharchives", (b, settings) => settings.SearchArchives = b },
					{ "uniquelines", (b, settings) => settings.UniqueLines = b },
					{ "verbose", (b, settings) => settings.Verbose = b },
					{ "version", (b, settings) => settings.PrintVersion = b },
				};

		public List<SearchOption> Options { get; private set; }
		public Dictionary<string, SearchOption> ArgDictionary { get; private set; }
		public Dictionary<string, SearchOption> FlagDictionary { get; private set; }

		public SearchOptions()
		{
			_searchOptionsResource = EmbeddedResource.GetResourceFileContents("CsSearch.Resources.searchoptions.xml");
			Options = new List<SearchOption>();
			ArgDictionary = new Dictionary<string, SearchOption>();
			FlagDictionary = new Dictionary<string, SearchOption>();
			SetOptionsFromXml();
		}

		private void SetOptionsFromXml()
		{
			var doc = XDocument.Parse(_searchOptionsResource);
			foreach (var f in doc.Descendants("searchoption"))
			{
				var longArg = f.Attributes("long").First().Value;
				var shortArg = f.Attributes("short").First().Value;
				var desc = f.Value.Trim();
				if (ArgActionDictionary.ContainsKey(longArg))
				{
					var option = new SearchArgOption(shortArg, longArg, ArgActionDictionary[longArg], desc);
					Options.Add(option);
					ArgDictionary.Add(longArg, option);
					if (!string.IsNullOrWhiteSpace(shortArg))
					{
						ArgDictionary.Add(shortArg, option);
					}
				}
				else if (BoolFlagActionDictionary.ContainsKey(longArg))
				{
					var option = new SearchFlagOption(shortArg, longArg, BoolFlagActionDictionary[longArg], desc);
					Options.Add(option);
					FlagDictionary.Add(longArg, option);
					if (!string.IsNullOrWhiteSpace(shortArg))
					{
						FlagDictionary.Add(shortArg, option);
					}
				}
			}
		}

		public SearchSettings SettingsFromArgs(IEnumerable<string> args)
		{
			// default to PrintResults = true since this is called from CLI functionality
			var settings = new SearchSettings {PrintResults = true};
			var queue = new Queue<string>(args);

			while (queue.Count > 0)
			{
				var s = queue.Dequeue();
				if (s.StartsWith("-"))
				{
					try
					{
						while (s.StartsWith("-"))
						{
							s = s.Substring(1);
						}
					}
					catch (InvalidOperationException e)
					{
						throw new SearchException(e.Message);
					}
					if (string.IsNullOrWhiteSpace(s))
					{
						throw new SearchException("Invalid option: -");
					}
					if (ArgDictionary.ContainsKey(s))
					{
						try
						{
							((SearchArgOption)ArgDictionary[s]).Action(queue.Dequeue(), settings);
						}
						catch (InvalidOperationException e)
						{
							throw new SearchException(e.Message);
						}
					}
					else if (FlagDictionary.ContainsKey(s))
					{
						try
						{
							((SearchFlagOption)FlagDictionary[s]).Action(true, settings);
						}
						catch (InvalidOperationException e)
						{
							throw new SearchException(e.Message);
						}
					}
					else
					{
						throw new SearchException("Invalid option: " + s);
					}
				}
				else
				{
					settings.StartPath = s;
				}
			}
			return settings;
		}

		public void Usage(int exitCode = 0)
		{
			Console.WriteLine(GetUsageString());
			Environment.Exit(exitCode);
		}

		public string GetUsageString()
		{
			var sb = new StringBuilder();
			sb.AppendLine("\nUsage:");
			sb.AppendLine(" CsSearch.exe [options] -s <searchpattern> <startpath>\n");
			sb.AppendLine("Options:");
			var optStrings = new List<string>();
			var optDescs = new List<string>();
			var longest = 0;
			foreach (var opt in Options.OrderBy(o => o.SortArg))
			{
				var optString = new StringBuilder();
				if (!string.IsNullOrWhiteSpace(opt.ShortArg))
				{
					optString.Append("-" + opt.ShortArg + ",");
				}
				optString.Append("--" + opt.LongArg);
				if (optString.Length > longest)
				{
					longest = optString.Length;
				}
				optStrings.Add(optString.ToString());
				optDescs.Add(opt.Description);
			}
			var format = " {0,-"+longest+"}  {1}";
			for (var i = 0; i < optStrings.Count; i++)
			{
				sb.AppendLine(string.Format(format, optStrings[i], optDescs[i]));
			}
			return sb.ToString();
		}
	}
}

File: node/nodesearch/searchoptions.js
/*
 * searchoptions.js
 *
 * defines the set of search options and provides functionality to define search settings from them
 */

var FileUtil = require('./fileutil.js').FileUtil;
var SearchOption = require('./searchoption.js').SearchOption;
var SearchSettings = require('./searchsettings.js').SearchSettings;

function SearchOptions() {
    var that = this;

    //TODO: move to config file
    var searchOptionsPath = '~/src/git/xsearch/shared/searchoptions.xml';

    // the list of SearchOption objects (populated by setOptionsFromXml)
    var options = [];
    var argMap = {};
    var flagMap = {};

    var argActionMap = {
        'in-archiveext':
            function(x, settings) { settings.addInArchiveExtension(x); },
        'in-archivefilepattern':
            function(x, settings) { settings.addInArchiveFilePattern(x); },
        'in-dirpattern':
            function(x, settings) { settings.addInDirPattern(x); },
        'in-ext':
            function(x, settings) { settings.addInExtension(x); },
        'in-filepattern':
            function(x, settings) { settings.addInFilePattern(x); },
        'in-linesafterpattern':
            function(x, settings) { settings.addInLinesAfterPattern(x); },
        'in-linesbeforepattern':
            function(x, settings) { settings.addInLinesBeforePattern(x); },
        'linesafter':
            function(x, settings) { settings.linesAfter = x; },
        'linesaftertopattern':
            function(x, settings) { settings.addInLinesAfterToPattern(x); },
        'linesafteruntilpattern':
            function(x, settings) { settings.addInLinesAfterUntilPattern(x); },
        'linesbefore':
            function(x, settings) { settings.linesBefore = x; },
        'maxlinelength':
            function(x, settings) { settings.maxLineLength = x; },
        'out-dirpattern':
            function(x, settings) { settings.addOutDirPattern(x); },
        'out-archiveext':
            function(x, settings) { settings.addOutArchiveExtension(x); },
        'out-archivefilepattern':
            function(x, settings) { settings.addOutArchiveFilePattern(x); },
        'out-ext':
            function(x, settings) { settings.addOutExtension(x); },
        'out-filepattern':
            function(x, settings) { settings.addOutFilePattern(x); },
        'out-linesafterpattern':
            function(x, settings) { settings.addOutLinesAfterPattern(x); },
        'out-linesbeforepattern':
            function(x, settings) { settings.addOutLinesBeforePattern(x); },
        'search':
            function(x, settings) { settings.addSearchPattern(x); }
    };

    var flagActionMap = {
        'allmatches':
            function(settings) { settings.firstMatch = false; },
        'archivesonly':
            function(settings) { settings.archivesOnly = true;
                                 settings.searchArchives = true; },
        'debug':
            function(settings) { settings.debug = true; },
        'dotiming':
            function(settings) { settings.doTiming = true; },
        'excludehidden':
            function(settings) { settings.excludeHidden = true; },
        'firstmatch':
            function(settings) { settings.firstMatch = true; },
        'includehidden':
            function(settings) { settings.excludeHidden = false; },
        'help':
            function(settings) { settings.printUsage = true; },
        'listdirs':
            function(settings) { settings.listDirs = true; },
        'listfiles':
            function(settings) { settings.listFiles = true; },
        'listlines':
            function(settings) { settings.listLines = true; },
        'multilinesearch':
            function(settings) { settings.multilineSearch = true; },
        'noprintmatches':
            function(settings) { settings.printResults = false; },
        'norecursive':
            function(settings) { settings.recursive = false; },
        'nosearcharchives':
            function(settings) { settings.searchArchives = false; },
        'printmatches':
            function(settings) { settings.printResults = true; },
        'recursive':
            function(settings) { settings.recursive = true; },
        'searcharchives':
            function(settings) { settings.searchArchives = true; },
        'uniquelines':
            function(settings) { settings.uniqueLines = true; },
        'verbose':
            function(settings) { settings.verbose = true; },
        'version':
            function(settings) { settings.printVersion = true; }
    };

    function optcmp(o1, o2) {
        var a = o1.sortarg;
        var b = o2.sortarg;
        return a.localeCompare(b);
    }

    // setOptionsFromXml
    (function () {
        var fs = require('fs');
        var DomJS = require('dom-js').DomJS;

        var domjs = new DomJS();

        var xml = fs.readFileSync(FileUtil.expandPath(searchOptionsPath)).toString();
        domjs.parse(xml, function(err, dom) {
            if (err) {
                throw err;
            }
            for (var i in dom.children) {
                var child = dom.children[i];
                if (child.name && child.name === 'searchoption') {
                    var longArg = child.attributes.long;
                    var shortArg = child.attributes.short;
                    var desc = child.text().trim();
                    var func = null;
                    if (argActionMap[longArg]) func = argActionMap[longArg];
                    else if (flagActionMap[longArg]) func = flagActionMap[longArg];
                    else throw new Error("Unknown option: "+longArg);
                    var option = new SearchOption(shortArg, longArg, desc, func);
                    options.push(option);
                    if (argActionMap[longArg]) {
                        argMap[longArg] = option;
                        if (shortArg) argMap[shortArg] = option;
                    } else if (flagActionMap[longArg]) {
                        flagMap[longArg] = option;
                        if (shortArg) flagMap[shortArg] = option;
                    }
                }
            }
        });
        options.sort(optcmp);
    })();

    this.searchSettingsFromArgs = function (args, callback) {
        var err = null;
        var settings = new SearchSettings();
        // default printResults to true since it's being run from cmd line
        settings.printResults = true;
        while(args) {
            var arg = args.shift();
            if (!arg) {
                break;
            }
            if (arg.charAt(0) === '-') {
                while (arg && arg.charAt(0) === '-') {
                    arg = arg.substring(1);
                }
                if (argMap[arg]) {
                    if (args) {
                        argMap[arg].func(args.shift(), settings);
                    } else {
                        err = new Error("Missing argument for option "+arg);
                    }
                } else if (flagMap[arg]) {
                    flagMap[arg].func(settings);
                    if (['h','help','V','version'].indexOf(arg) > -1)
                        return settings;
                } else {
                    err = new Error("Unknown option: "+arg);
                }
            } else {
                settings.startPath = arg;
            }
        }
        if (settings.debug) {
            settings.verbose = true;
        }
        callback(err, settings);
    };

    this.usage = function () {
        usageWithCode(0);
    };

    this.usageWithCode = function (exitCode) {
        console.log(getUsageString());
        process.exit(exitCode);
    };

    var getUsageString = function () {
        var usage = 'Usage:\nnodesearch [options] <startpath>\n\n';
        usage += 'Options:\n';
        var optStrings = [];
        var optDescs = [];
        var longest = 0;
        for (var o in options) {
            var opt = options[o];
            var optString = '';
            if (opt.shortarg)
                optString += '-' + opt.shortarg + ',';
            optString += '--' + opt.longarg;
            if (optString.length > longest)
                longest = optString.length;
            optStrings.push(optString);
            optDescs.push(opt.desc)
        }
        for (var o in optStrings) {
            var optString = optStrings[o];
            while (optString.length < longest)
                optString += ' ';
            usage += optString + '  ' + optDescs[o] + '\n';
        }
        return usage;
    };
}

exports.SearchOptions = SearchOptions;
